src/providers/ipa/ipa_subdomains_ext_groups.c

```diff
@@ -326,62 +334,168 @@ static errno_t add_ad_user_to_cached_groups(struct ldb_dn *user_dn,
+    }
+
+    /* Add all new IPA groups to SYSDB_ORIG_MEMBEROF because they are most
+     * probalby remove by the previous user update and mark all new groups as
+     * processed. */
     for (c = 0; groups[c] != NULL; c++) {
-        if (groups[c][0] == '\0') {
-            continue;
+        ret = sysdb_attrs_add_string(user_attrs, SYSDB_ORIG_MEMBEROF,
+                                     groups[c]);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "sysdb_attrs_add_string failed.\n");
+            goto done;
         }
 
-        ret = sysdb_search_groups_by_orig_dn(tmp_ctx, group_dom, groups[c],
+        groups[c][0] = '\0';
+    }
+
+    if (DEBUG_IS_SET(SSSDBG_TRACE_ALL)) {
+        for (c = 0; sysdb_ipa_group_memberships[c] != NULL; c++);
+        DEBUG(SSSDBG_TRACE_ALL, "Cached IPA groups [%zu].\n", c);
+
+        for (c = 0; groups[c] != NULL; c++);
+        DEBUG(SSSDBG_TRACE_ALL, "New IPA groups [%zu].\n", c);
+
+        for (c = 0; add_groups[c] != NULL; c++);
+        DEBUG(SSSDBG_TRACE_ALL, "Groups to add [%zu].\n", c);
+
+        for (c = 0; del_groups[c] != NULL; c++);
+        DEBUG(SSSDBG_TRACE_ALL, "Groups to delete [%zu].\n", c);
+    }
+
+    /* TODO: there is a similar functionality (adding and removing group
+     * memberships in sysdb_update_members_ex(), but the missing group feature
+     * is missing. It might be worth to evaluate if either the missing group
+     * feature can be added there or if group which are missing in the cache
+     * can bew handled differently here. */
+
+    for (c = 0; add_groups[c] != NULL; c++) {
+
+        ret = sysdb_search_groups_by_orig_dn(tmp_ctx, group_dom, add_groups[c],
                                              NULL, &msgs_count, &msgs);
         if (ret != EOK) {
             if (ret == ENOENT) {
                 DEBUG(SSSDBG_TRACE_ALL, "Group [%s] not in the cache.\n",
-                                         groups[c]);
+                                         add_groups[c]);
                 *missing_groups = true;
+                talloc_free(groups[d]);
+                /* add missing group back to the list */
+                groups[d++] = talloc_steal(groups, add_groups[c]);
                 continue;
             } else {
                 DEBUG(SSSDBG_OP_FAILURE, "sysdb_search_entry failed.\n");
                 goto done;
             }
         }
 
-/* TODO? Do we have to remove members as well? I think not because the AD
- * query before removes all memberships. */
-
         ret = sysdb_mod_group_member(group_dom, user_dn, msgs[0]->dn,
                                      LDB_FLAG_MOD_ADD);
         if (ret != EOK && ret != EEXIST) {
             DEBUG(SSSDBG_OP_FAILURE, "sysdb_mod_group_member failed.\n");
             goto done;
         }
 
-        user_attrs = sysdb_new_attrs(tmp_ctx);
-        if (user_attrs == NULL) {
-            DEBUG(SSSDBG_OP_FAILURE, "sysdb_new_attrs failed.\n");
-            ret = ENOMEM;
+    }
```

Comment:
Isn' t a `NULL` terminator needed?

Reply:
`groups` is already `NULL` terminated, but adding it to the end of the missing list might help to avoid a couple of unneeded comparisons. Added.